File: 1886.determine-whether-matrix-can-be-obtained-by-rotation.py
#
# @lc app=leetcode id=1886 lang=python3
#
# [1886] Determine Whether Matrix Can Be Obtained By Rotation
#

# @lc code=start
class Solution:
    def findRotation(self, mat: List[List[int]], target: List[List[int]]) -> bool:
        def rotate(mat):
            n = len(mat)
            for i in range(n // 2):
                for j in range(i, n - i - 1):
                    (mat[i][j], mat[j][n - i - 1], mat[n - i - 1][n - j - 1], mat[n - j - 1][i],) = (
                        mat[n - j - 1][i],
                        mat[i][j],
                        mat[j][n - i - 1],
                        mat[n - i - 1][n - j - 1],
                    )

        for _ in range(4):
            if mat == target:
                return True
            rotate(mat)
        return False


# @lc code=end

File: RustSolutions/src/1886.determine-whether-matrix-can-be-obtained-by-rotation.rs
/*
 * @lc app=leetcode id=1886 lang=rust
 *
 * [1886] Determine Whether Matrix Can Be Obtained By Rotation
 */

// @lc code=start
impl Solution {
    pub fn find_rotation(mut mat: Vec<Vec<i32>>, target: Vec<Vec<i32>>) -> bool {
        let n = mat.len();

        if mat == target {
            return true;
        }
        (0..3).any(|_| {
            for (i, j) in (0..(n / 2)).flat_map(|i| (i..(n - i - 1)).map(move |j| (i, j))) {
                let (tmp, b, r) = (mat[i][j], n - j - 1, n - i - 1);
                mat[i][j] = mat[b][i];
                mat[b][i] = mat[r][b];
                mat[r][b] = mat[j][r];
                mat[j][r] = tmp;
            }
            mat == target
        })
    }
}
// @lc code=end
